packages/babel-types/src/definitions/core.js

```diff
@@ -291,7 +294,11 @@ export const functionCommon = {
     optional: true,
   },
   typeParameters: {
-    validate: assertNodeType("TypeParameterDeclaration", "Noop"),
+    validate: assertNodeType(
```

Comment:
Guess one thing to think about, is that this change will make `tSDeclareMethod` and `tSDeclareFunction` (which spread `functionCommon` accept `TypeParameterDeclaration`.

Reply:
Good catch -- I missed that. Would you rather override the property for the TS types, or compose more granularly?

```js
defineType("TSDeclareFunction", {
  ...functionCommon,
  returnType {
    ...functionCommon.returnType
    validate: ...etc
  },
  typeParameters: {
    ...functionCommon.typeParameters,
    validate: ...etc.
  }
});
```
vs.
```js
export const functionCommon = {
  // everything there now but returnType and typeAnnotation
}

// only used for core types that need to accept either Flow or TS nodes
const functionCommonWithTypes = {
  returnType: {},
  typeAnnotation: {}
}
```
I didn't see other examples of overriding spread properties in definintions -- if that's generally the case, then the second approach might be more consistent with the rest of the code.

Reply:
Yeah let's go with the second

Reply:
@existentialism It's not immediately obvious because I squashed the commit, but this has been addressed.

I split the type annotations out of `functionCommon` into `functionTypeAnnotationCommon`, and spread both objects on the relevant core/es2015 types.

`functionDeclarationCommon` and `classMethodOrDeclareMethodCommon` still spread just  `functionCommon`. Types that use those now also spread either `functionTypeAnnotationCommon` or `tSFunctionTypeAnnotationCommon` as appropriate.